MCO-593: Consume MCO API and CRDs from openshift/api

jkyros opened the pull request:
This just brings: 

- The API from https://github.com/openshift/api/pull/1453
- The clients from https://github.com/openshift/client-go/pull/243

back into the MCO. 

This is almost entirely find/replace except: 
- API helpers moved into apihelpers package (since they couldn't go with the API) 
- existing API removed from the MCO 
- hack scripts updated to no longer generate deepcopy code 
- hack scripts updated to copy MCO CRDs out of vendor directory when "make update" is run

jkyros commented:
/test e2e-gcp-op

jkyros commented:
/test e2e-aws-ovn-upgrade

jkyros commented:
/test all

jkyros commented:
/test verify

jkyros commented:
/test all

jkyros commented:
/test unit 
/test verify

jkyros commented:
/test e2e-gcp-op

jkyros commented:
/test e2e-gcp-op

jkyros commented:
/retest-required

cdoern commented:
@jkyros I *think* you need to modify the codegen to also fetch the `operator.openshift.io` group so we can generate listers for the generic CRD. I have been working on this as well in my branch :) The zz_generated... files in pkg/api need to be modified by hand as well I think

jkyros commented:
/test all

jkyros commented:
/test e2e-aws-ovn-upgrade

jkyros commented:
Oh, hah, we're failing the upgrade test because master still has those certificate API fields as strings. We changed them to metav1.Time, but I guess this test upgrades from "whatever is in master" (rather than 4.13) to "whatever is in this PR", so we'd need to fix that in master, and then re-run this  : 
```
2023-08-15T05:03:40.989575279Z W0815 05:03:40.989527       1 reflector.go:533] github.com/openshift/client-go/machineconfiguration/informers/externalversions/factory.go:101: failed to list *v1.MachineConfigPool: parsing time "2033-08-12 01:47:54 +0000 UTC" as "2006-01-02T15:04:05Z07:00": cannot parse " 01:47:54 +0000 UTC" as "T"
2023-08-15T05:03:40.989575279Z E0815 05:03:40.989555       1 reflector.go:148] github.com/openshift/client-go/machineconfiguration/informers/externalversions/factory.go:101: Failed to watch *v1.MachineConfigPool: failed to list *v1.MachineConfigPool: parsing time "2033-08-12 01:47:54 +0000 UTC" as "2006-01-02T15:04:05Z07:00": cannot parse " 01:47:54 +0000 UTC" as "T"
2023-08-15T05:04:05.304139210Z W0815 05:04:05.304088       1 reflector.go:533] github.com/openshift/client-go/machineconfiguration/informers/externalversions/factory.go:101: failed to list *v1.ControllerConfig: parsing time "2033-08-12 01:47:54 +0000 UTC" as "2006-01-02T15:04:05Z07:00": cannot parse " 01:47:54 +0000 UTC" as "T"
2023-08-15T05:04:05.304139210Z E0815 05:04:05.304121       1 reflector.go:148] github.com/openshift/client-go/machineconfiguration/informers/externalversions/factory.go:101: Failed to watch *v1.ControllerConfig: failed to list *v1.ControllerConfig: parsing time "2033-08-12 01:47:54 +0000 UTC" as "2006-01-02T15:04:05Z07:00": cannot parse " 01:47:54 +0000 UTC" as "T"
```
Everything else is green though, so I think we've got it aside from that.

jkyros commented:
@rioliu-rh this is passing all the tests except for `e2e-aws-ovn-upgrade `  (or at least it was before I pulled it out of draft), and it's not going to pass that one because:
-  the API review required us to [make a breaking change](https://github.com/openshift/api/pull/1453#discussion_r1288134482) (vs master) to our API, and 
- that test seemingly upgrades from "whatever is in master" to "whatever is in this PR" and they're quite incompatible


A 4.13 upgrade to 4.14 with "this PR" upgrade still works. 

 I don't really have anything specific that's changing here that needs to be tested, as this should more or less be "the same stuff that was in the MCO before", but from a new place and generated by a newer generator. If it didn't line up right, it would be outright broken. but if you have a chance to check for regressions, that might not be terrible. 

The only thing I could really see breaking it is something like server-side apply with our required CRD fields --  since the new generated clients use the applyconfiguration library functions under the hood, and we previously didn't have applyconfiguration functions at all. 

( I will obviously take the replace directive out of go.mod before this merges, but I can't until the api + clients PRs merge)

jkyros commented:
/test okd-scos-images
/test e2e-hypershift

jkyros commented:
/test e2e-hypershift

jkyros commented on Makefile:
```diff
@@ -101,7 +100,7 @@ install-tools: install-golangci-lint install-setup-envtest install-go-junit-repo
 verify: install-tools
 	./hack/golangci-lint.sh $(GOTAGS)
-	hack/verify-codegen.sh
+#	hack/verify-codegen.sh
```
It's a long story. :smile: 

- Upstream deprecated generate-groups and (among other things) removed execute permission from `generate-internal-groups.sh`
- If I fix that, what it's actually doing is effectively a no-op now that the api packages are gone. Verify verifies by re-generating and then comparing what it generated to what exists, but because nothing else we have generates anything, nothing exists and pkg/generated is empty: 
```
WARNING: generate-groups.sh is deprecated.
WARNING: Please use k8s.io/code-generator/kube_codegen.sh instead.

WARNING: generate-internal-groups.sh is deprecated.
WARNING: Please use k8s.io/code-generator/kube_codegen.sh instead.

Generating clientset for machineconfiguration.openshift.io:v1 at github.com/openshift/machine-config-operator/pkg/generated/clientset
Generating listers for machineconfiguration.openshift.io:v1 at github.com/openshift/machine-config-operator/pkg/generated/listers
Generating informers for machineconfiguration.openshift.io:v1 at github.com/openshift/machine-config-operator/pkg/generated/informers
+ echo 'diffing hack/../pkg against freshly generated codegen'
diffing hack/../pkg against freshly generated codegen
+ ret=0
+ diff -Naupr hack/../_tmp/pkg hack/../pkg
+ cp -R hack/../_tmp/pkg/apihelpers hack/../_tmp/pkg/constants hack/../_tmp/pkg/controller hack/../_tmp/pkg/daemon hack/../_tmp/pkg/operator hack/../_tmp/pkg/server hack/../_tmp/pkg/version hack/../pkg
+ [[ 0 -eq 0 ]]
+ echo 'hack/../pkg up to date.'
hack/../pkg up to date.
+ cleanup
+ rm -rf hack/../_tmp
```
So it's successfully comparing nothing to nothing, which would only be a useful verification step in the event someone accidentally added a types.go somewhere. And even then it would just be like "generated doesn't match exsiting, please updating existing" which is probably not what we want unless we think we're going to start stuffing CRD types back in here again. 

To actually do what I think we'd want (verify that our codegen was properly generated from our API), we'd probably need to duplicate the verification happening in openshift/api and openshift/client-go, but they already verify it? So I don't know if it would be worth duplicating that here?

yuqi-zhang replied:
Ah, thanks for the context! I think that makes sense, I'm fine with removing it then

jkyros commented on install/0000_80_machine-config-operator_01_containerruntimeconfig.crd.yaml:
```diff
@@ -5,6 +5,7 @@ metadata:
   labels:
     "openshift.io/operator-managed": ""
   annotations:
+    api-approved.openshift.io: https://github.com/openshift/api/pull/1453
```
That was just me trying to reduce the blast radius and not having thought through all the collateral damage.

Other operators do something similar, e.g. https://github.com/openshift/machine-api-operator/blob/master/hack/crds-sync.sh but I don't think anything stops us from just having the Dockerfile copy them out of vendor and into /manifests when it builds. I'll try it. :smile:

sinnykumari commented on hack/crds-sync.sh:
```diff
@@ -0,0 +1,32 @@
+cp "vendor/github.com/openshift/api/machineconfiguration/v1/0000_80_controllerconfig.crd.yaml" "manifests/controllerconfig.crd.yaml"
+
+
+#v1/0000_10_containerruntimeconfig.crd.yaml:0000_80_machine-config-operator_01_containerruntimeconfig.crd.yaml
```
Wondering what is the purposed of below commented lines?

jkyros replied:
John miiiight have left some of his "scratch work" in the script :smile:

sinnykumari commented on hack/crds-sync.sh:
```diff
@@ -0,0 +1,32 @@
+               "v1/0000_80_kubeletconfig.crd.yaml:0000_80_machine-config-operator_01_kubeletconfig.crd.yaml"
+               "v1/0000_80_machineconfig.crd.yaml:0000_80_machine-config-operator_01_machineconfig.crd.yaml"
+               "v1/0000_80_machineconfigpool.crd.yaml:0000_80_machine-config-operator_01_machineconfigpool.crd.yaml" ) 
+                #TODO(jkyros): 0000_80_machine-config-operator_02_containerruntimeconfig.crd.yaml)
```
Is there anything blocking us to copy this from vendor/ ?

jkyros replied:
Nah, I just hadn't tried it yet. Looks like it works, I took this whole script out.

sinnykumari commented:
/payload 4.15 nightly blocking

jkyros commented:
Alright, cleaned this up a bit per featureGate and CRD sync discussions:
   - Added some commit descriptions :smile:  
   - Did one last round of manual `kubelet.yaml` featureGate sync, we will do it the right way in a follow-up 
   - CRD sync script has been removed 
   - Dockerfile now copies CRDs straight from /vendor (It doesn't look like this broke any of our hack scripts, but I might be missing something). Yes, I did regenerate Dockerfile.rhel7. 
   - added library-go bump to my second/OpenShift go-deps bump to match https://github.com/openshift/machine-config-operator/pull/3958

jkyros commented:
kuryr issue appears to be fixed, lets see if I broke it: 
/payload 4.15 nightly blocking

yuqi-zhang commented:
/payload 4.15 nightly blocking

looks like some jobs got killed? Just to be safe, running again

jkyros commented:
Hmm, so of those failures: 

**1. periodic-ci-openshift-release-master-nightly-4.15-e2e-aws-sdn-upgrade**:

Failed, but it looks like all for different reasons? Well, and consistently the network-node-identity webhook pod had problems: 
```
{  event [ns/openshift-network-node-identity pod/network-node-identity-sjnmg node/ip-10-0-56-210.us-west-2.compute.internal hmsg/8ce90c027c - pathological/true interesting/true reason/BackOff Back-off restarting failed container webhook in pod network-node-identity-sjnmg_openshift-network-node-identity(6c3cc642-a4ad-4f90-95dc-3b66497e0693)] happened 24 times
event [ns/openshift-network-node-identity pod/network-node-identity-2qr7g node/ip-10-0-44-151.us-west-2.compute.internal hmsg/faa0915261 - pathological/true interesting/true reason/BackOff Back-off restarting failed container webhook in pod network-node-identity-2qr7g_openshift-network-node-identity(916a1e2b-9c39-44f1-8202-75da7edad0e0)] happened 25 times
event [ns/openshift-network-node-identity pod/network-node-identity-2qr7g node/ip-10-0-44-151.us-west-2.compute.internal hmsg/faa0915261 - pathological/true interesting/true reason/BackOff Back-off restarting failed container webhook in pod network-node-identity-2qr7g_openshift-network-node-identity(916a1e2b-9c39-44f1-8202-75da7edad0e0)] happened 47 times
```
I'm thinking probably not us. 

**2. periodic-ci-openshift-hypershift-release-4.15-periodics-e2e-aws-ovn-conformance**:

Good ol' Hypershift :smile:. looks maybe concerning because it's consistent failure:  
``` 
     STEP: Destroying namespace "e2e-test-ccm-tqcck" for this suite. @ 10/11/23 23:08:34.954
  â€¢ [FAILED] [0.799 seconds]
  [sig-cloud-provider][Feature:OpenShiftCloudControllerManager][Late] [It] Deploy an external cloud provider [apigroup:machineconfiguration.openshift.io] [Suite:openshift/conformance/parallel]
  github.com/openshift/origin/test/extended/cloud_controller_manager/ccm.go:23

    [FAILED] Expected
        <bool>: false
    to be true
    In [It] at: github.com/openshift/origin/test/extended/cloud_controller_manager/ccm.go:35 @ 10/11/23 23:08:34.927
  ------------------------------

  Summarizing 1 Failure:
    [FAIL] [sig-cloud-provider][Feature:OpenShiftCloudControllerManager][Late] [It] Deploy an external cloud provider [apigroup:machineconfiguration.openshift.io] [Suite:openshift/conformance/parallel]
    github.com/openshift/origin/test/extended/cloud_controller_manager/ccm.go:35

  Ran 1 of 1 Specs in 0.799 seconds
  FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
```
but it looks like it's failing because the cloud-controller-manager pods are missing, and so far I can't find enough data in the gathers to figure out why that might be. I also don't know how this would prevent those pods from being present. 

Like, this is before the MCO/kubelet controller even does anything with it: https://github.com/openshift/origin/blob/a12dbde9ad9d82e06c74fccc8eefa22b966696a5/test/extended/cloud_controller_manager/ccm.go#L35 


**3. periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-gcp-ovn-rt-upgrade**:

At least some of those runs passed so that looks like maybe flakes? And the baremetal operator making too many watches?
```
{  fail [github.com/openshift/origin/test/extended/apiserver/api_requests.go:449]: Expected     <[]string \| len:1, cap:1>: [         "Operator \"cluster-baremetal-operator\" produces more watch requests than expected: watchrequestcount=98, upperbound=88, ratio=1.11",     ] to be empty Ginkgo exit error 1: exit with code 1}

```

jkyros commented:
I wonder if I can run just the hypershift one
/payload-job periodic-ci-openshift-hypershift-release-4.15-periodics-e2e-aws-ovn-conformance

sinnykumari commented:
Overall LGTM, we will need to look at failing HyperShift conformance job

jkyros commented:
Testing in https://github.com/openshift/machine-config-operator/pull/3976 it looks like it's actually the "CRDs from vendor" commit that breaks it. I might have changed the operator timing by removing the CRD sync from the operator somehow, or Hypershift cares that controllerconfig is part of the payload now for some reason?  ( both possibilities are weird, but this seems like a weird thing )

If not having this in is blocking stuff, I can peel that CRD commit of and then we can troubleshoot that separately?

jkyros commented:
As soon as this run fails, I'll peel the commit off (I just want to rule out a problem being fixed in the background): 
/payload-job periodic-ci-openshift-hypershift-release-4.15-periodics-e2e-aws-ovn-conformance

cdoern commented:
@jkyros looks like it failed?

yuqi-zhang commented:
/payload-job periodic-ci-openshift-hypershift-release-4.15-periodics-e2e-aws-ovn-conformance

Just another data point to see if we missed anything

jkyros commented:
cluster failure, and a pod sandbox failure 
/test e2e-aws-ovn-upgrade
/test okd-scos-e2e-aws-ovn

jkyros commented:
/test e2e-openstack

jkyros commented:
Alright, ci passed, hypershift conformance passed. I'll let this run overnight: 
/payload 4.15 nightly blocking

jkyros commented:
Oh wow that payload test was super green. :smile: 

The Azure one was the only one that failed, and it passed twice, failed twice. Failures were: 

```
{  etcdHighCommitDurations was at or above info for at least 1m58s on platformidentification.JobType{Release:"4.15", FromRelease:"4.15", Platform:"azure", Architecture:"amd64", Network:"ovn", Topology:"ha"} (maxAllowed=0s): pending for 0s, firing for 1m58s:
```
which I'm reasonably confident this didn't affect.

cdoern commented:
/lgtm
/hold cancel
